3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,225 @@
+                  'Discarding event message received prior to new creative' +
+                  ' message for' + creativeId);
+              this.creativeMessageRouters_[creativeId]
+                  .sendMessagesToListener(messages);
+            } catch (e) {
+              dev().error(TAG_, 'Failed to send message to listener: ' +
+                  e.message);
+            }
+          });
+        });
+    this.iframeMessagingClient_.sendMessage(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.READY);
+  }
+}
+
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative. Also sends response messages from the iframe meant for this
+ * particular creative.
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!Window} win The enclosing window object
+   * @param {!IframeMessagingClient} iframeMessagingClient Facilitates
+   * communication with the frame
+   * @param {!string} creativeId The ID of the creative to route messages
+   * to/from
+   * @param {string=} opt_extraData Extra data to be passed to the frame
+   */
+  constructor(win, iframeMessagingClient, creativeId, opt_extraData) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @private {!string} */
+    this.creativeId_ = creativeId;
+
+    /** @private {?string} */
+    this.extraData_ = opt_extraData;
+
+    /** @private {?function(!Array<!AmpAnalytics3pEvent>)} */
+    this.eventListener_ = null;
+
+    if (this.win_.onNewAmpAnalyticsInstance) {
+      try {
+        this.win_.onNewAmpAnalyticsInstance(this);
+      } catch (e) {
+        dev().error(TAG_, 'Exception thrown by onNewAmpAnalyticsInstance in' +
+          this.win_.location.href + ': ' + e.message);
+      }
+    } else {
+      dev().error(TAG_, 'Must implement onNewAmpAnalyticsInstance in' +
+        this.win_.location.href);
+    }
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur.
+   * There may only be one listener. If another function has previously been
+   * registered as a listener, it will no longer receive events.
+   * @param {!function(!Array<AmpAnalytics3pEvent>)} listener A function
+   * that takes an array of event strings, and does something with them.
+   */
+  registerAmpAnalytics3pEventsListener(listener) {
+    if (this.eventListener_) {
+      dev().warn(TAG_, 'Replacing existing eventListener for ' +
+        this.creativeId_);
+    }
+    this.eventListener_ = listener;
+  }
+
+  /**
+   * Receives message(s) from a creative for the cross-domain iframe
+   * and passes them to that iframe's listener, if a listener has been
+   * registered
+   * @param {!Array<AmpAnalytics3pEvent>} messages The message that was received
+   */
+  sendMessagesToListener(messages) {
+    if (!messages.length || !this.eventListener_) {
```

Comment:
Given the fix to your previous comment, it is no longer possible to reach here without implementing onNewAmpAnalyticsInstance, as the enclosing AmpAnalytics3pMessageRouter will now not have been created.
However, this could still happen if they have done the first line here but not the second:
```
      window.onNewAmpAnalyticsInstance = ampAnalytics => {
        ampAnalytics.registerAmpAnalytics3pEventsListener(events => {
```
Unfortunately we can't go and create that for them, so I just made the dev().warn message explicitly say "Attempted to send messages when no listener  configured. Be sure to first call  registerAmpAnalytics3pEventsListener()"